### PR TITLE
Do not remove mpirun when OpenMPI is built with Slurm

### PR DESCRIPTION
**Pull request by JiaweiZhuang**
#7574 removed the `mpirun`/`mpiexec` executable when OpenMPI is built with slurm support. I suggest restoring them for a few reasons:

1.  Calling `mpirun` after `salloc`/`sbatch` is a documented practice on OpenMPI website. See https://www.open-mpi.org/faq/?category=slurm. People might follow the instruction and get surprised -- either not find `mpirun` , or an incorrect version like `/usr/bin/mpirun` is called. Indeed this is how I find this issue.
2. Directly using `mpirun` is useful for fine-tuning the MCA parameter, e.g. force TCP and turn off IB & Vader by `mpirun --mca btl tcp,self`. Useful for debugging sometimes :) . With`srun` it can still be done by setting `OMPI_MCA_<param_name>`, but a bit inconvenient (https://www.open-mpi.org/faq/?category=tuning#setting-mca-params). Using `orterun` feels exotic for most users.
3. In certain cases `mpirun` can have better performance than `srun` (though probably rare with `srun --mpi=pmix`). See https://www.mail-archive.com/users@lists.open-mpi.org/msg31874.html

Just a small suggestion :)  In any case, I really appreciate #7574 (and #8427) -- OpenMPI + slurm works quite well so far!

### CODE REVIEW

**Comment by citibeth** [comment]
@alalazo Do you have opinions on this?  What would be the downside of NOT  removing the mpirun/mpiexec executable when OpenMPI is built with slurm support.

**Comment by alalazo** [comment]
**TL;DR** The issues we had are well explained in the e-mail you linked. Removing links prevents people from using `mpirun` instead of `srun` accidentally (and experience severely degraded performance). As far as our experience goes, we didn't notice real downsides removing `mpirun` and `mpiexec`, just benefits.

--- 
 The main issue we experienced when people were using `mpirun` with `slurm` is wrong CPU binding. From the same [mailing list thread]( https://www.mail-archive.com/users@lists.open-mpi.org/msg31874.html) you posted:

> So beware of resource requests that specify only the number of tasks 
(—ntasks=64) and then launch with “mpiexec”.  Slurm will happily allocate those 
tasks anywhere it can (on a busy cluster) and you will get some very 
non-optimal core mappings/bindings and, possibly, core sharing.

We experienced inefficiencies along the same lines as above - tasks migrating from one core to the other, multiple tasks mapped to the same core etc. My take on that is:

1. Both OpenMPI and Slurm want "by design" to be in charge of process spawning and task layout in ways that could conflict with each other.
2. The fact that you installed Slurm is a hint that you trust it

In the end this change eliminated completely for our site (EPFL) an entire class of user tickets: those complaining about severely degraded performance when using OpenMPI. 

Also worth mentioning if you are managing a site, removing `mpirun` and `mpiexec` solved the sporadic reports of newbies running a few jobs on the front-end (just because they were used to `mpirun` on their laptop and tried to give the same commands once logged on the clusters).

A few comments on the points you make above:

> 1. [...] Calling mpirun after salloc/sbatch is a documented practice on OpenMPI website.

True, but usually there's site documentation to refer to + in the Slurm website they [suggest differently](https://slurm.schedmd.com/mpi_guide.html#open_mpi). I guess this goes back to the fact that both OpenMPI and Slurm want to be in charge of process spawning.

> 2. [...] Using orterun feels exotic for most users.

That was a deliberate choice, as my intent was to encourage the use of `srun` if OpenMPI was compiled with Slurm support. That's basically what they say in the thread you linked:

> Anyway, that is what we have observed.  Generally speaking, I try to get users 
to use “srun” but many of them still use “mpiexec” out of habit.  You know what 
they say about old habits.  

Removing the link was a way to remove "old habits".

**Comment by goxberry** [comment]
I appreciate hearing the reasoning for removing those commands, because I missed the rationale for their removal at my site.

I will point out also, however, that there’s some testing infrastructure in other projects that assumes `mpiexec` exists (e.g., `hypre` did this for a while, as did `petsc`, I think, but it’s been a while since I checked). It would be helpful, if nothing else, to have some way to run those testing scripts without having to change each of  them on a per-case basis. I don’t know how feasible it is to write an `mpirun` or `mpiexec` wrapper script for `srun`, but when I raised this issue with the LC, that’s what they told me to do. (I declined.)

**Comment by jrood-nrel** [comment]
I use OpenMPI with Slurm and MPICH with Slurm. I agree with @alalazo , but there are cases I would like to have `mpirun` available. I think this should be a variant since I consider it a personal preference. Maybe a `+mpirun` defaulted to on.

**Comment by citibeth** [comment]
Sounds like this should be added as a variant, with the default set to how
things currently are.

**Comment by jrood-nrel** [comment]
@citibeth I would add that the default is entirely unexpected behavior though.

**Comment by alalazo** [comment]
> I think most people wouldn't expect this.

That's the entire point of removing the links **when compiling with Slurm**. People expect `mpirun` or `mpiexec` to be there, then they use the wrappers and their jobs fail in inconvenient ways (your program runs, but much slower - or it segfaults but only sometimes etc.).  By removing `mpirun` we effectively prevented those issues while losing nothing else.

> Sounds like this should be added as a variant, with the default set to how
things currently are.

The variant is already there I think. Check `legacylaunchers`.

**Comment by JiaweiZhuang** [comment]
@alalazo Thanks for the detailed explanation. Also, just realized the variant `legacylaunchers`. Looks good to me. Closing this PR.

**Comment by sethrj** [comment]
@alalazo @JiaweiZhuang I just caught up on this thread (and the chat on slack). Perhaps an easy fix to the "surprise" aspect for Spack users would be (when `~legacylaunchers`) to install a tiny script aliased to `mpirun`/`mpi`:
```sh
#!/bin/sh
echo "This version of Spack is installed without the mpiexec/mpirun commands to prevent"
echo "unintended performance issues. See github.com/.... for a discussion."
echo "If you understand the potential consequences of a misconfigured `mpirun`, you can"
echo "use spack to install `openmpi+legacylaunchers` to restore the executables."
exit 2
```

**Comment by alalazo** [comment]
@sethrj Looks reasonable to me. Can you to submit a PR?

**Comment by gmfricke** [comment]
This seems to breaks spack's Orca build by default.

**Comment by hppritcha** [comment]
This issue will need to be reopened for the 5.0 OMPI release.  Starting with that release,  only PMIx will be supported for interaction with the job launcher.  Many slurm installations don't even bother building the PMIx plugin, so that means no direct srun launch support in many cases, and no  real support for launching at all, unless mpirun is installed.

@alalazo

**Comment by alalazo** [comment]
@hppritcha Would it be a solution to rename the variant to something more meaningful, and give it a "positive" meaning? We can call the variant `+mpirun` and describe it saying:
```
If False delete the `mpirun` and `mpiexec` wrappers, useful for Slurm with OpenMPI < v5.0 
```
and then add:
```
conflicts('~mpirun', when=`@5.0:`, msg='Starting from OpenMPI v5.0 wrappers are required')
```
Would this be sufficient?
